src/main/kotlin/com/terraformation/backend/tracking/db/ObservationStore.kt

```diff
@@ -626,6 +626,36 @@ class ObservationStore(
     }
   }
 
+  /**
+   * Fetches last completed observation for a planting site if there are no other upcoming or in
+   * progress observations, otherwise returns most recently updated observation.
+   */
+  fun fetchLastCompletedObservationTime(plantingSiteId: PlantingSiteId): Instant? {
+    return dslContext
+        .select(DSL.max(OBSERVATIONS.COMPLETED_TIME))
+        .from(OBSERVATIONS)
+        .where(OBSERVATIONS.PLANTING_SITE_ID.eq(plantingSiteId))
+        .andNotExists(
+            DSL.selectOne()
+                .from(OBSERVATIONS)
+                .where(OBSERVATIONS.PLANTING_SITE_ID.eq(plantingSiteId))
+                .and(
+                    OBSERVATIONS.STATE_ID.`in`(
+                        ObservationState.InProgress,
+                        ObservationState.Overdue,
+                        ObservationState.Upcoming)))
+        .fetchOne(DSL.max(OBSERVATIONS.COMPLETED_TIME))
+  }
+
+  fun hasObservations(plantingSiteId: PlantingSiteId): Boolean {
+    return dslContext
+        .selectOne()
+        .from(OBSERVATIONS)
+        .where(OBSERVATIONS.PLANTING_SITE_ID.eq(plantingSiteId))
+        .fetch()
+        .isNotEmpty
+  }
```

Comment:
This will return a row for each of the observations for the site; one option would be to add `.limit(1)`.

However, jOOQ has a convenience function for this pattern that takes a table name and a varargs list of conditions. I only just recently noticed it myself, but it's perfect for this:
```kotlin
return dslContext
    .fetchExists(
        OBSERVATIONS,
        OBSERVATIONS.PLANTING_SITE_ID.eq(plantingSiteId))
```

Reply:
This is great, will use `fetchExists` for these types of checks.